src/libraries/System.Net.Sockets/src/System/Net/Connections/SocketsConnectionFactory.cs

```diff
@@ -0,0 +1,159 @@
+
+            Socket socket = CreateSocket(_addressFamily, _socketType, _protocolType, endPoint, options);
+
+            try
+            {
+                using var args = new TaskSocketAsyncEventArgs();
+                args.RemoteEndPoint = endPoint;
+
+                if (socket.ConnectAsync(args))
+                {
+                    using (cancellationToken.UnsafeRegister(o => Socket.CancelConnectAsync((SocketAsyncEventArgs)o!), args))
+                    {
+                        await args.Task.ConfigureAwait(false);
+                    }
+                }
+
+                if (args.SocketError != SocketError.Success)
+                {
+                    SocketException ex = new SocketException((int)args.SocketError);
+                    if (args.SocketError == SocketError.OperationAborted && cancellationToken.IsCancellationRequested)
+                    {
+                        throw new TaskCanceledException(s_cancellationMessage, ex, cancellationToken);
+                    }
+
+                    throw NetworkErrorHelper.MapSocketException(ex);
+                }
+
+                return new SocketConnection(socket, this, options);
+            }
+            catch (SocketException socketException)
+            {
+                socket.Dispose();
+                throw NetworkErrorHelper.MapSocketException(socketException);
+            }
+            catch
+            {
+                socket.Dispose();
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Creates the socket that shall be used with the connection.
+        /// </summary>
+        /// <param name="addressFamily">The <see cref="AddressFamily"/> to forward to the socket.</param>
+        /// <param name="socketType">The <see cref="SocketType"/> to forward to the socket.</param>
+        /// <param name="protocolType">The <see cref="ProtocolType"/> to forward to the socket.</param>
+        /// <param name="endPoint">The <see cref="EndPoint"/> this socket will be connected to.</param>
+        /// <param name="options">Properties, if any, that might change how the socket is initialized.</param>
+        /// <returns>A new unconnected <see cref="Socket"/>.</returns>
+        /// <remarks>
+        /// In case of TCP sockets, the default implementation of this method will create a socket with <see cref="Socket.NoDelay"/> enabled.
+        /// In case of IPv6 sockets <see cref="Socket.DualMode"/> is also be enabled.
+        /// </remarks>
+        protected virtual Socket CreateSocket(
+            AddressFamily addressFamily,
+            SocketType socketType,
+            ProtocolType protocolType,
+            EndPoint? endPoint,
+            IConnectionProperties? options)
+        {
+            Socket socket = new Socket(addressFamily, socketType, protocolType);
+
+            if (protocolType == ProtocolType.Tcp)
+            {
+                socket.NoDelay = true;
```

Comment:
Nice. Creating a SocketsConnectionFactory for UDS shouldn't require a subclass anymore 👍

Reply:
If we don't already have one, we need a test for UDS added.  We had 7-ish scenarios outlined for HttpClient extensibility; I'd really like to see us have a test for each, to help validate the design actually accomplishes what we want it to.

Reply:
We have it:
https://github.com/dotnet/runtime/blob/edb8332a3dac3b59a62a715babeb760ea91a5499/src/libraries/System.Net.Sockets/tests/FunctionalTests/Connections/SocketsConnectionFactoryTests.cs#L19

https://github.com/dotnet/runtime/blob/edb8332a3dac3b59a62a715babeb760ea91a5499/src/libraries/System.Net.Sockets/tests/FunctionalTests/Connections/SocketsConnectionFactoryTests.cs#L111-L132

Reply:
> Do we also want to test it with HttpClient ?

Yes.  We need end-to-end tests showing that the scenario actually works.  That means using HttpClient to access an endpoint over a Unix domain socket, using HttpClient with a bandwidth monitoring extension, using HttpClient with a connection factory that connects to all resolved addresses in parallel and picks the first that succeeds, using HttpClient with a socket bound to a particular network interface, etc.

Reply:
In case of questions, who has the best knowledge on the details here, who I can ping on Monday? @JamesNK? @scalablecory ?

Reply:
(to be clear, it needn't be in this PR. I'm commenting in general about the whole feature)

Reply:
@antonfirsov see "Requests for HttpClient extensibility" at https://github.com/dotnet/runtime/issues/1793

Reply:
I think the amount of remaining work deserves a new issue. Opened #40624 so we can discuss and track it.